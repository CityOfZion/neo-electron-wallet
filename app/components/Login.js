import React from 'react';
import { connect } from 'react-redux';
import { login, setBalance, resetPrice, getAccounts } from '../actions/index.js';
import { Link } from 'react-router';
import CreateWallet from './CreateWallet.js'
import { getBlockByIndex } from '../wallet/api.js';
import { addAccountToLocalStorage, clearLocalStorage } from '../wallet/index.js'
import SavedWallets from './SavedWallets.js'

let input_wif;

const onWifChange = (dispatch) => {
  // lookup wif address to check whether it is valid and enable login
  dispatch(login(input_wif.value));
  // somewhat hacky way to reset price on every login
  // dispatch(resetPrice());

};

let Login = ({ dispatch, loggedIn, wif }) =>
  <div id="loginPage">
    <div className="login">
      <input type="text" placeholder="Enter your private key here (WIF)" onChange={() => onWifChange(dispatch)} ref={node => {input_wif = node;}} />
      <div className="loginButtons">
        {loggedIn ? <button><Link to="/dashboard">Login</Link></button> : <button disabled="true">Login</button>}
        <button><Link to="/create">New Wallet</Link></button>
        <button onClick={()=> clearLocalStorage()}>Clear Storage</button>
        <SavedWallets />
      </div>
    </div>
  </div>;

const mapStateToProps = (state) => ({
  loggedIn: state.account.loggedIn,
  wif: state.account.wif
});

Login = connect(mapStateToProps)(Login);

export default Login;
