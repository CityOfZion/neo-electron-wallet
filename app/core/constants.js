// @flow
export const NETWORK = {
  MAIN: 'MainNet',
  TEST: 'TestNet'
}

export const EXPLORER = {
  NEO_TRACKER: 'Neotracker',
  ANT_CHAIN: 'Antchain'
}

export const ASSETS = {
  NEO: 'NEO',
  GAS: 'GAS'
}

export const ASSETS_LABELS = {
  [ASSETS.NEO]: 'Neo',
  [ASSETS.GAS]: 'Gas'
}

export const BIP44_PATH =
  '8000002C' +
  '80000378' +
  '80000000' +
  '00000000' +
  '00000000'
