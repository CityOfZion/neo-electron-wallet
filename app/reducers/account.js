import * as types from '../actions/types';
import { getAccountsFromWIFKey } from '../wallet/index.js';

// reducer that manages account state (account now = private key)
export default (state = {wif: null, address:null, loggedIn: false}, action) => {
  switch (action.type) {
    case types.LOGIN:
      let loadAccount;
      try {
        loadAccount = getAccountsFromWIFKey(action.wif)[0];
      }
      catch (e){ loadAccount = -1; }
      if(loadAccount === -1 || loadAccount === -2 || loadAccount === undefined){
        return {...state, wif:action.wif,  loggedIn:false};
      }
      return {...state, wif:action.wif, address:loadAccount.address, loggedIn:true};
    case types.LOGOUT:
      return {'wif': null, address: null, 'loggedIn': false};
    default:
      return state;
  }
};