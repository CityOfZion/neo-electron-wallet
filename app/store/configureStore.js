import { createStore, applyMiddleware } from 'redux';
import { createLogger } from 'redux-logger';
import thunk from 'redux-thunk';
import rootReducer from '../reducers/rootReducer';
// export type GlobalState = {}
export var INITIAL_STATE = {};
var loggerMiddleware = createLogger();
function configureStore(initialState) {
    if (initialState === void 0) { initialState = INITIAL_STATE; }
    return createStore(rootReducer, initialState, process.env.NODE_ENV === 'production'
        ? applyMiddleware(thunk)
        : applyMiddleware(thunk, loggerMiddleware));
}
export var store = configureStore();
