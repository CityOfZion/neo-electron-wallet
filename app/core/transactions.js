// @flow
/* eslint-disable camelcase */
import { ASSETS, ASSET_HASHES } from './constants'

export const parseTransactions = (neoscanTxs) => {
  return neoscanTxs
    .map(({ txid, asset_moved, amount_moved, balance, block_height }) => {
      let NEO = 0
      let GAS = 0
      let gas_sent = false
      let neo_sent = false
      let other_sent = false

      if (ASSET_HASHES[asset_moved] === ASSETS.NEO) {
        neo_sent = true
      } else if (ASSET_HASHES[asset_moved] === ASSETS.GAS) {
        gas_sent = true
      } else {
        // TODO handle additional assets for both APIs in the future
      }

      balance.forEach(({ asset, amount }) => {
        if (asset === ASSETS.NEO) NEO = amount
        if (asset === ASSETS.GAS) GAS = amount
      })

      return {
        txid,
        block_index: block_height,
        GAS,
        NEO,
        gas_sent,
        neo_sent
      }
    })
}
