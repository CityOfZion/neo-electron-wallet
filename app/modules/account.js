// @flow
import { getAccountFromWIFKey } from 'neon-js'
import { verifyPrivateKey } from '../core/wallet'
import { sendEvent, clearTransactionEvent } from './transactions'

// Constants
export const LOGIN = 'LOGIN'
export const LOGOUT = 'LOGOUT'
export const SET_DECRYPTING = 'SET_DECRYPTING'
export const SET_KEYS = 'SET_KEYS'

// Actions
export function login (wif: string) {
  return {
    type: LOGIN,
    wif: wif
  }
}

export function logout () {
  return {
    type: LOGOUT
  }
}

export function decrypting (bool: Boolean) {
  return {
    type: SET_DECRYPTING,
    state: bool
  }
}

export function setKeys (keys: any) {
  return {
    type: SET_KEYS,
    keys
  }
}

export const onWifChange = (history: Object, wif: string) => (dispatch: DispatchType) => {
  if (verifyPrivateKey(wif)) {
    dispatch(login(wif))
    history.push('/dashboard')
  } else {
    dispatch(sendEvent(false, 'That is not a valid private key'))
    setTimeout(() => dispatch(clearTransactionEvent()), 5000)
  }
}

// Reducer that manages account state (account now = private key)
export default (state: Object = {wif: null, address: null, loggedIn: false, redirectUrl: null, decrypting: false, accountKeys: []}, action: Object) => {
  switch (action.type) {
    case LOGIN:
      let loadAccount: Object | number
      try {
        loadAccount = getAccountFromWIFKey(action.wif)
      } catch (e) { loadAccount = -1 }
      if (loadAccount === -1 || loadAccount === -2 || loadAccount === undefined) {
        return {...state, wif: action.wif, loggedIn: false}
      }
      // $FlowFixMe
      return {...state, wif: action.wif, address: loadAccount.address, loggedIn: true, decrypting: false}
    case LOGOUT:
      return {...state, 'wif': null, address: null, 'loggedIn': false, decrypting: false}
    case SET_DECRYPTING:
      return {...state, decrypting: action.state}
    case SET_KEYS:
      return {...state, accountKeys: action.keys}
    default:
      return state
  }
}
