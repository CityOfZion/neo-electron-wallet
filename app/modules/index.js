// @flow
import { combineReducers } from 'redux'
import api from './api'
import generateWallet from './generateWallet'
import transactions from './transactions'
import wallet from './wallet'
import claim from './claim'
import notifications from './notifications'
import modal from './modal'
import addressBook from './addressBook'

export default combineReducers({
  api,
  generateWallet,
  wallet,
  transactions,
  claim,
  notifications,
  modal,
  addressBook
})
