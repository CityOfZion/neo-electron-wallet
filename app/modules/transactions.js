// @flow
/* eslint-disable camelcase */
import { api } from 'neon-js'

import { setTransactionHistory, getNEO, getGAS, getTokens, getScriptHashForNetwork } from './wallet'
import { showErrorNotification, showInfoNotification, showSuccessNotification } from './notifications'
import { getWif, getPublicKey, getSigningFunction, getAddress, LOGOUT } from './account'
import { getNetwork } from './metadata'

import { validateTransactionBeforeSending, obtainTokenBalance, isToken } from '../core/wallet'
import { ASSETS } from '../core/constants'
import asyncWrap from '../core/asyncHelper'

import { log } from '../util/Logs'

// Constants
export const LOADING_TRANSACTIONS = 'LOADING_TRANSACTIONS'

export const setIsLoadingTransaction = (isLoading: boolean) => ({
  type: LOADING_TRANSACTIONS,
  payload: {
    isLoadingTransactions: isLoading
  }
})

export const syncTransactionHistory = (net: NetworkType, address: string) => async (dispatch: DispatchType) => {
  dispatch(setIsLoadingTransaction(true))
  const [err, transactions] = await asyncWrap(api.neonDB.getTransactionHistory(net, address))
  if (!err && transactions) {
    const txs = transactions.map(({ NEO, GAS, txid, block_index, neo_sent, neo_gas }: TransactionHistoryType) => ({
      type: neo_sent ? ASSETS.NEO : ASSETS.GAS,
      amount: neo_sent ? NEO : GAS,
      txid,
      block_index
    }))
    dispatch(setIsLoadingTransaction(false))
    dispatch(setTransactionHistory(txs))
  } else {
    dispatch(setIsLoadingTransaction(false))
  }
}

export const sendTransaction = (sendAddress: string, sendAmount: string, symbol: TokenSymbolType) => async (dispatch: DispatchType, getState: GetStateType): Promise<*> => {
  const state = getState()
  const wif = getWif(state)
  const address = getAddress(state)
  const net = getNetwork(state)
  const neo = getNEO(state)
  const gas = getGAS(state)
  const tokens = getTokens(state)
  const signingFunction = getSigningFunction(state)
  const publicKey = getPublicKey(state)

  const rejectTransaction = (message: string) => dispatch(showErrorNotification({ message }))
  const tokenBalance = isToken(symbol) && obtainTokenBalance(tokens, symbol)
  const parsedSendAmount = parseFloat(sendAmount)

  const { error, valid } = validateTransactionBeforeSending(neo, gas, tokenBalance, symbol, sendAddress, parsedSendAmount)
  if (valid) {
    const selfAddress = address
    let sendAsset = {}
    sendAsset[symbol] = parseFloat(parsedSendAmount)

    dispatch(showInfoNotification({ message: 'Sending Transaction...', autoDismiss: 0 }))
    log(net, 'SEND', selfAddress, { to: sendAddress, asset: symbol, amount: parsedSendAmount })

    const isHardwareSend = !!publicKey

    let sendAssetFn
    if (isHardwareSend) {
      dispatch(showInfoNotification({ message: 'Please sign the transaction on your hardware device', autoDismiss: 0 }))
      sendAssetFn = () => api.neonDB.doSendAsset(net, sendAddress, publicKey, sendAsset, signingFunction)
    } else {
      if (symbol === ASSETS.NEO || symbol === ASSETS.GAS) {
        sendAssetFn = () => api.neonDB.doSendAsset(net, sendAddress, wif, sendAsset, null)
      } else {
        const scriptHash = getScriptHashForNetwork(net, symbol)
        const decimalAdjustedSendAmount = parsedSendAmount * 10 ** tokens[symbol].info.decimals
        sendAssetFn = () => api.nep5.doTransferToken(net, scriptHash, wif, sendAddress, decimalAdjustedSendAmount)
      }
    }

    const [err, response] = await asyncWrap(sendAssetFn())
    if (err || response.result === undefined || response.result === false) {
      return rejectTransaction('Transaction failed!')
    } else {
      return dispatch(showSuccessNotification({ message: 'Transaction complete! Your balance will automatically update when the blockchain has processed it.' }))
    }
  } else {
    return rejectTransaction(error)
  }
}

// state getters
export const getIsLoadingTransactions = (state: Object) => state.transactions.isLoadingTransactions

const initialState = {
  isLoadingTransactions: false
}

export default (state: Object = initialState, action: ReduxAction) => {
  switch (action.type) {
    case LOADING_TRANSACTIONS:
      const { isLoadingTransactions } = action.payload
      return {
        ...state,
        isLoadingTransactions
      }
    case LOGOUT:
      return initialState
    default:
      return state
  }
}
