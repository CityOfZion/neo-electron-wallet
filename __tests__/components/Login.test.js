import React from 'react'
import configureStore from 'redux-mock-store'
import { shallow, mount } from 'enzyme'

import { login } from '../../app/modules/account'

import { sendEvent } from '../../app/modules/transactions'
import { KEYS } from '../../app/core/constants'

import Login from '../../app/containers/LoginPrivateKey/LoginPrivateKey'

const setup = (state = { account: {
  loggedIn: true,
  wif: undefined
}}) => {
  const store = configureStore()(state)
  const wrapper = shallow(<Login store={store} />)

  return {
    store,
    wrapper
  }
}

const state = {
  account: {
    loggedIn: false,
    wif: null
  }
}

describe('Login', () => {
  test('renders without crashing', () => {
    const { wrapper } = setup()
    expect(wrapper).toMatchSnapshot()
  })

  test('clicking show key icon toggles private key visibility', () => {
    const { wrapper } = setup()

    expect(wrapper.state('showKey')).toEqual(false)

    wrapper
      .find('.viewKey')
      .simulate('click')

    expect(wrapper.state('showKey')).toEqual(true)
  })

  test('pressing enter on passphrase input field triggers sendEvent', () => {
    const store = configureStore()(state)
    const wrapper = mount(<Login store={store} />)
    const c = wrapper.find('input')

    // console.log('debug:'+c.debug())
    // console.log('debug2:'+c.html())
    c.simulate('keyDown', { key: KEYS.ENTER })
    expect(store.getActions()[0]).toEqual(sendEvent(false, 'That is not a valid private key'))
  })

  test('private key field input onChange dispatches LOGIN action', () => {
    const history = {
      push: jest.fn()
    }

    const store = configureStore()(state)
    const wrapper = mount(<Login store={store} history={history} />)
    const c = wrapper.find('input')

    c.instance().value = 'L1xpshXfzF6iQTq42onA5km8qwyzBaNQzPADhfTt2jzzcQSVoP5A'
    c.simulate('change')
    wrapper.find('.loginButton').simulate('click')

    expect(store.getActions()[0]).toEqual(login('L1xpshXfzF6iQTq42onA5km8qwyzBaNQzPADhfTt2jzzcQSVoP5A'))
  })

  // test('private key field input onChange dispatches LOGIN action', () => {
  //   const { store, wrapper } = setup();
  //   let preventDefault = jest.fn();
  //   const deeperWrapper = wrapper.shallow();
  //
  //   const someWif = 'L1xpshXfzF6iQTq42onA5km8qwyzBaNQzPADhfTt2jzzcQSVoP5A'
  //   deeperWrapper
  //     .find('input')
  //     .simulate('change', { target: { value: someWif } })
  //     .find('button')
  //     .simulate('click');
  //
  //   expect(store.getActions()).toEqual([
  //     { wif: someWif, type: 'LOGIN' }
  //   ]);
  // });
})
