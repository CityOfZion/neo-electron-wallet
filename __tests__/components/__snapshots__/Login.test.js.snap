// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login renders without crashing 1`] = `
<div
  id="loginPage"
>
  <div
    className="login"
  >
    <div
      className="loginForm"
    >
      <Logo />
      <input
        autoFocus={true}
        onChange={[Function]}
        placeholder="Enter your private key here (WIF)"
        type="password"
      />
      <FaEye
        className="viewKey"
        onClick={[Function]}
      />
    </div>
    <div
      className="loginButtons"
    >
      <button
        onClick={[Function]}
      >
        Login
      </button>
      <Link
        onlyActiveOnIndex={false}
        style={Object {}}
        to="/"
      >
        <button
          className="altButton"
        >
          Home
        </button>
      </Link>
    </div>
    <Footer />
  </div>
</div>
`;
